Baremetal deploy (#5139)
redwoodjs/redwood: docs/docs/deploy/baremetal.md

With Redwood's Baremetal deployment option, the source (like your dev machine) will SSH into one or more remote machines and execute commands in order to update your codebase, run any database migrations and restart services.

> Deploying from a client (like your own development machine) consists of running a single command.

Reviewer: I'd keep the colon here, I use it to say "the thing I just described to you I'm about to show on the following lines"

> Deploying from a client (like your own development machine) consists of running a single command.

Reviewer: I'd keep the colon here, I use it to say "the thing I just described to you I'm about to show on the following lines"

> First time deploy

Reviewer: We've got sort of an informal standard that we keep an empty newline between text and the following codeblocks

> First time deploy

Reviewer: ```suggestion
First time deploy

```

Empty newline before codeblocks

> First time deploy

Reviewer: ```suggestion
First time deploy

```

Empty newline before codeblocks

```bash
yarn rw deploy baremetal --first run
```

> Subsequent deploys

Reviewer: Newline

> Subsequent deploys

Reviewer: ```suggestion
Subsequent deploys

```

Empty newline before codeblocks

> Subsequent deploys

Reviewer: ```suggestion
Subsequent deploys

```

Empty newline before codeblocks

```bash
yarn rw deploy baremetal
```

> ## Deployment lifecycle

Reviewer: ```suggestion
## Deployment Lifecycle
```

Headings should be Title Case

> ## Deployment lifecycle

Reviewer: ```suggestion
## Deployment Lifecycle
```

Headings should be Title Case

3. `yarn rw prisma generate` - generates latest Prisma Client libs
4. `yarn rw dataMigrate up` - runs data migrations, ignoring them if not installed
5. `yarn rw build` - builds the web and/or api sides

> 6. `yarn pm2 restart [service]` - restarts the serving process(es).

Reviewer: When we've got a bullet list like this we either put a period on all of them or none of them

> ### First run lifecycle

Reviewer: Empty newline after headings

> If the `--first-run` flag is specified step 6. aboce will be skipped and the following commands will run instead:

Reviewer: ```suggestion
If the `--first-run` flag is specified step 6. above will be skipped and the following commands will run instead:
```

> If the `--first-run` flag is specified step 6. aboce will be skipped and the following commands will run instead:

Reviewer: ```suggestion
If the `--first-run` flag is specified step 6. above will be skipped and the following commands will run instead:
```
```suggestion
If the `--first-run` flag is specified step 6. aboce will be skipped and the following commands will run instead:
```

  - `yarn pm2 start [service]` - starts the serving process(es)

>   - `yarn pm2 save` - saves the running services to the deploy users config file for future startup. See `First Deploy - Starting on Reboot` for further information

Reviewer: Instead of making `First Deploy - Starting on Reboot` a code block, you can just leave it plain text but make it a link to the section below

Author: completed

>   - `yarn pm2 save` - saves the running services to the deploy users config file for future startup. See `First Deploy - Starting on Reboot` for further information

Reviewer: Instead of making `First Deploy - Starting on Reboot` a code block, you can just leave it plain text but make it a link to the section below

If there are any issues the deploy should stop and you'll see the error message printed to the console. Assume it worked, hooray! You're deployed to BAREMETAL.

> ### Starting on reboot

Reviewer: ```suggestion
### Starting on Reboot

```

Title Case and empty newline after heading

> ### Starting on reboot

Reviewer: ```suggestion
### Starting on Reboot

```

Title Case and empty newline after heading

The `pm2` service requires some system "hooks" to be installed so it can boot up using your systems service manager.  Otherwise, your services will need to be manually started again on reboot.  These steps only need to be run the first time you deploy to a machine.

> 1. Ssh into your server as you did for the "Server Setup".  Navigate to your source folder.  For example `cd /var/www/example`

Reviewer: I tend to make "SSH" either all caps, or all lowercase (like the command itself). Looks like I used a mix of both in this doc, but we should really settle on one or the other. Probably all caps like the man page does: https://linux.die.net/man/1/ssh

Author: updating this

> 1. Ssh into your server as you did for the "Server Setup".  Navigate to your source folder.  For example `cd /var/www/example`

Reviewer: I tend to make "SSH" either all caps, or all lowercase (like the command itself). Looks like I used a mix of both in this doc, but we should really settle on one or the other. Probably all caps like the man page does: https://linux.die.net/man/1/ssh

> 2. Run the command `yarn pm2 startup`.  You will see some output similar to the output below.  Select this output starting with `sudo` and run it.  *Note* you will be using sudo so will need the root password to complete this

Reviewer: I'd put `sudo` in a code block here so that it looks like a command

> 2. Run the command `yarn pm2 startup`.  You will see some output similar to the output below.  Select this output starting with `sudo` and run it.  *Note* you will be using sudo so will need the root password to complete this

Reviewer: ```suggestion
2. Run the command `yarn pm2 startup`.  You will see some output similar to the output below. See the output after "copy/paste the following command:"? You'll need to do just that: copy the command starting with `sudo` and then paste and execute it. *Note* this command uses `sudo` so you'll need the root password to the machine in order for it to complete successfully.
```

I had to re-read that section of a couple of times to get what it was saying...I think this makes it a little clearer (I also included a little section below that explains it again with the actual command).

Author: Thank you.  It's good to have a second set of eyes on this for clarity!

> 2. Run the command `yarn pm2 startup`.  You will see some output similar to the output below.  Select this output starting with `sudo` and run it.  *Note* you will be using sudo so will need the root password to complete this

Reviewer: ```suggestion
2. Run the command `yarn pm2 startup`.  You will see some output similar to the output below.  Select this output starting with `sudo` and run it.  *Note* you will be using `sudo` so will need the root password to complete this
```

> The below text is an *example* output.  Yours will be different

```bash

> deploy@redwood:/var/www/my-app$ yarn pm2 startup

Reviewer: You mentioned above that this command would require `sudo` but it's not using `sudo` here

> deploy@redwood:/var/www/my-app$ yarn pm2 startup

Reviewer: You mentioned above that this command would require `sudo` but it's not using `sudo` here

[PM2] Init System found: systemd
[PM2] To setup the Startup Script, copy/paste the following command:
sudo env PATH=$PATH:/home/deploy/.nvm/versions/node/v17.8.0/bin /var/www/my-app/node_modules/pm2/bin/pm2 startup systemd -u deploy --hp /home/deploy
```

> 

Reviewer: ```suggestion

In this example, you would copy `sudo env PATH=$PATH:/home/deploy/.nvm/versions/node/v17.8.0/bin /var/www/my-app/node_modules/pm2/bin/pm2 startup systemd -u deploy --hp /home/deploy` and run it.

```
